# Module 01: Python Basics

**By the end of this notebook you will be able to:**
- store values in variables and know their type
- do arithmetic and build text messages
- make decisions with `if` / `elif` / `else`
- repeat work with a `for` loop
- package steps into a reusable function

Run every cell (Shift + Enter). Then change a value and run it again, that is how it sticks.

## 1. Variables: labeled boxes for values
A **variable** is a name that holds a value. Think of it as a labeled box. You put something in, and use the label later. We use `=` to assign.

In [2]:
bill = 16.99
tip = 1.01
print(bill)
print(tip)

16.99
1.01


## 2. Types: what kind of value is it?
Every value has a **type**. The three you'll use constantly:
- `int` — whole number (2, 100)
- `float` — decimal number (16.99)
- `str` — text, in quotes ("Sunday")
- `bool` — True or False

`type()` tells you which one you have.

In [3]:
print(type(bill))      # float
print(type(2))         # int
print(type("Sunday"))  # str
print(type(True))      # bool

<class 'float'>
<class 'int'>
<class 'str'>
<class 'bool'>


Why care? Because analytics is all about types. A column of numbers (`float`) you can average; a column of labels (`str`) you count. Picking the right tool starts with knowing the type, exactly what we do in this course.

## 3. Arithmetic
The usual math works. Note `**` is "to the power of".

In [4]:
total = bill + tip
print(total)
print(bill * 0.15)   # a 15% tip would be...
print(2 ** 3)        # 2 to the power of 3

18.0
2.5484999999999998
8


## 4. Strings and f-strings
Text is a `str`. To drop a value into a sentence, put an `f` before the quotes and wrap the variable in `{ }`. This is an **f-string**, and you'll use it to label outputs.

In [5]:
day = "Sunday"
print(f"The bill was ${bill} on {day}.")

The bill was $16.99 on Sunday.


## 5. Booleans and comparisons
A comparison asks a yes/no question and returns a `bool` (`True`/`False`). These power every filter and every `if`.

In [6]:
print(bill > 20)      # is the bill over 20?
print(tip == 1.01)    # == means "is equal to"
print(bill != 10)     # != means "is not equal to"

False
True
True


Note: `=` *assigns* a value, but `==` *compares*. Mixing them up is the single most common beginner bug.

## 6. Making decisions: if / elif / else
Run different code depending on a condition. The indented lines belong to their branch.

In [7]:
bill = 16.99

if bill > 30:
    print("Big bill")
elif bill > 15:
    print("Medium bill")
else:
    print("Small bill")

Medium bill


Python reads top to bottom and takes the **first** branch whose condition is True. Change `bill` above and rerun to see different branches fire.

## 7. Repeating work: the for loop
A `for` loop runs the same code once for each item in a collection. Here we react to several bills without copying code five times.

In [8]:
bills = [16.99, 10.34, 21.01, 23.68, 24.59]

for b in bills:
    print(f"Bill: ${b}")

Bill: $16.99
Bill: $10.34
Bill: $21.01
Bill: $23.68
Bill: $24.59


`b` takes each value in turn. `range(n)` is handy when you just want to repeat n times or count.

In [9]:
for i in range(5):   # 0, 1, 2, 3, 4
    print(i)

0
1
2
3
4


## 8. Functions: package steps to reuse
A **function** bundles steps under a name so you can run them again with different inputs. `def` defines it; `return` hands back the result.

In [ ]:
def tip_amount(bill, percent):
    return bill * (percent / 100)

print(tip_amount(16.99, 15))   # 15% of 16.99
print(tip_amount(50, 20))      # 20% of 50

One definition, reused for any bill and any percent. That's the whole point of a function: write it once, use it everywhere.

## Recap
You can now use variables and types, do arithmetic, build messages with f-strings, compare values, branch with `if`, repeat with `for`, and define functions. That's the backbone of every program you'll write this term.

## Your challenge
Open `practice.ipynb` and try the exercises. First one: write a function that takes a bill and returns the total including a 18% tip.